### PR TITLE
Fixes #13177 - Adds proxy_status as base class, adds tftp status class

### PR DESCRIPTION
**Pull request by shlomizadok**

### CODE REVIEW

**Comment by shlomizadok** [comment]
@domcleal - though it is set as wip, I'd love to hear your comments.
The goal of this change is to:
- Extend ProxyStatus so other plugins can hook into and add their own connectivity (so for instance puppet will have a PuppetProxyStatus, etc.,)
- Remove inflation of SmartProxy model, with status methods.

Thanks :smile:

File: app/models/smart_proxy.rb [hunk 1]
```diff
@@ -91,7 +89,7 @@ def has_feature?(feature)
   private
 
   def proxy_status
-    @proxy_status ||= ProxyStatus.new(self)
+    @proxy_status ||= ProxyStatus::Base.new(self)
```
Comment (shlomizadok) [inline app/models/smart_proxy.rb]: This is kept here as it is needed to invoke features cache

**Comment by shlomizadok** [comment]
[test]

File: app/controllers/smart_proxies_controller.rb [hunk 2]
```diff
@@ -38,11 +38,11 @@ def refresh
   end
 
   def ping
-    requested_data(:version)
+    requested_data(ProxyStatus::Base.new(@smart_proxy), :version)
```
Comment (ohadlevy) [inline app/controllers/smart_proxies_controller.rb]: i think this should be converted to a before filter? then you would initialize it only once (in case you needed it another place later on)

File: app/services/proxy_status/base.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    private
+
+    attr_reader :proxy, :cache_duration, :api
+    alias_method :version, :api_versions
+
+    def fetch_proxy_data
+      begin
+        yield
+      rescue *ProxyStatus::CONNECTION_ERRORS => exception
+        raise ::Foreman::WrappedException.new exception, N_("Unable to connect to smart proxy")
```
Comment (ohadlevy) [inline app/services/proxy_status/base.rb]: I would consider changing this just to "unable to connect", imho its clear you are talking about a proxy.

File: app/services/proxy_status/tftp.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+module ProxyStatus
+  class TFTP < Base
+    def initialize(proxy, opts = {})
+      super(proxy, opts, 'TFTP')
+    end
+
+    def tftp_server
+      raise ::Foreman::Exception.new(N_('No TFTP feature for %s'), proxy.to_label) unless proxy.has_feature?('TFTP')
```
Comment (ohadlevy) [inline app/services/proxy_status/tftp.rb]: can this move to the base initializer?

File: app/models/smart_proxy.rb (hunk 1 above)
Comment (tbrisker) [inline app/models/smart_proxy.rb]: Just a thought - maybe make this method public, and have it return a hash of features => status objects. Then you can use for example `@smart_proxy.proxy_status[:tftp]` instead of `ProxyStatus::TFTP.new(@smart_proxy)` every time you want to get info from the status object? This method will create a status object for each feature that has one the first time it is called.

File: app/services/proxy_status/tftp.rb (hunk 4 above)
Comment (shlomizadok) [inline app/services/proxy_status/tftp.rb]: This would mean, I'd have to rescue on the controller action (where I call `.new`), which I'm less happy to do. I think that initializing should succeed and fail when calling the `tftp_server` method.
What do you think?

Comment (shlomizadok) [inline app/services/proxy_status/tftp.rb]: With latest change, this is a precaution of a precaution (e.g., if no tftp feature no chance it will get here...)

**Comment by shlomizadok** [comment]
[test]

File: app/models/smart_proxy.rb [hunk 5]
```diff
@@ -67,7 +65,7 @@ def self.smart_proxy_ids_for(hosts)
   end
 
   def refresh
-    proxy_status.revoke_cache!
+    proxy_status[:version].revoke_cache!
```
Comment (ohadlevy) [inline app/models/smart_proxy.rb]: should we revoke all caches instead of just the version?

File: app/models/smart_proxy.rb [hunk 6]
```diff
@@ -88,12 +86,19 @@ def has_feature?(feature)
     self.features.any? { |proxy_feature| proxy_feature.name == feature }
   end
 
-  private
-
   def proxy_status
-    @proxy_status ||= ProxyStatus.new(self)
+    @proxy_status ||= { :version => ProxyStatus::Version.new(self) }
+    features.map(&:name).each do |feature|
+      proxy_status_class = "ProxyStatus::#{feature.delete(' ')}".constantize rescue nil
```
Comment (ohadlevy) [inline app/models/smart_proxy.rb]: you can detect it also via the proxy status registry instead of trying to constantize it first.

Comment (ohadlevy) [inline app/models/smart_proxy.rb]: or just use directly something like:

``` ruby
status = ProxyStatus.find_status_by_humanized_name(name)
status.new(self) if status
...
```

File: app/services/proxy_status/tftp.rb (hunk 4 above)
Comment (ohadlevy) [inline app/services/proxy_status/tftp.rb]: meaning we can drop it?

File: app/services/proxy_status/version.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
+module ProxyStatus
+  class Version < Base
+    def initialize(proxy, opts = {})
+      super(proxy, opts, 'Version')
+    end
+
+    def api_versions
```
Comment (ohadlevy) [inline app/services/proxy_status/version.rb]: just versions?

File: app/controllers/smart_proxies_controller.rb [hunk 8]
```diff
@@ -63,8 +66,12 @@ def destroy
...
 
-  def requested_data(data_name)
-    data = @smart_proxy.public_send(data_name)
+  def find_status
+    @proxy_status = @smart_proxy.proxy_status
+  end
+
+  def requested_data(object, method)
+    data = object.public_send(method)
     render :json => {:success => true, :message => data}
```
Comment (tbrisker) [inline app/controllers/smart_proxies_controller.rb]: just a thought - this is basically a wrapper. So instead of passing an object and method, how about something along the lines of:

```
def requested_data
  render :json => {:success => true, :message => yield}
rescue Foreman::Exception => exception
  render :json => {:success => false, :message => exception.message} and return
end
```

Comment (tbrisker) [inline app/controllers/smart_proxies_controller.rb]: or even get rid of the hash and use HTTP response status to indicate success or failure?

File: app/models/smart_proxy.rb (hunk 6 above)
Comment (shlomizadok) [inline app/models/smart_proxy.rb]: The problem is that it is not registered until I initialize it, and here I try to initialize it.

Comment (ohadlevy) [inline app/models/smart_proxy.rb]: you could probably register it on when the code is parsed, i think thats
why @ares used require_dependency.

Comment (ares) [inline app/models/smart_proxy.rb]: the reason was that STI wouldn't work correctly but that seems to be right solution in here as well

File: app/controllers/smart_proxies_controller.rb [hunk 9]
```diff
@@ -38,11 +39,13 @@ def refresh
...
   def ping
-    requested_data(:version)
+    requested_data(@proxy_status[:version], :version)
   end
 
   def tftp_server
-    requested_data(:tftp_server)
+    @proxy_status[:tftp] ?
+      requested_data(@proxy_status[:tftp], :tftp_server) :
+      render(:json => {:success => false, :message => _('No TFTP feature')}) and return
```
Comment (ohadlevy) [inline app/controllers/smart_proxies_controller.rb]: nitpick: I don think the && return is needed here.

File: app/models/smart_proxy.rb [hunk 10]
```diff
@@ -88,12 +86,18 @@ def has_feature?(feature)
...
 
-  private
-
   def proxy_status
-    @proxy_status ||= ProxyStatus.new(self)
+    @proxy_status ||= { :version => ProxyStatus::Version.new(self) }
+    features.map(&:name).each do |feature|
+      normalized_feature_name = feature.delete(' ')
+      status = ProxyStatus.find_status_by_humanized_name(normalized_feature_name)
+      @proxy_status[normalized_feature_name.downcase.to_sym] = status.new(self) if status
```
Comment (ohadlevy) [inline app/models/smart_proxy.rb]: maybe do the downcasing above, so you'll be consistent?

Comment (ohadlevy) [inline app/models/smart_proxy.rb]: also maybe consider updating the feature validation? so only one feature with the same name (case insensitive ) can be created?

File: app/models/smart_proxy.rb [hunk 11]
```diff
@@ -88,12 +86,18 @@ def has_feature?(feature)
...
+    end
+    @proxy_status
```
Comment (ohadlevy) [inline app/models/smart_proxy.rb]: this will always create the proxy statues regardless if @proxy_status is set, is that what you wanted to do?

File: app/services/proxy_status.rb [hunk 12]
```diff
@@ -1,51 +1,14 @@
...
-      yield
-    rescue *CONNECTION_ERRORS => exception
-      raise ::Foreman::WrappedException.new exception, N_("Unable to connect to smart proxy")
-    end
+  def self.find_status_by_humanized_name(name)
+    status_registry.find { |s| s.humanized_name == name }
   end
 end
+require_dependency('proxy_status/version')
+require_dependency('proxy_status/tftp')
```
Comment (ohadlevy) [inline app/services/proxy_status.rb:14]: do you need to add one for each status? how do you achieve this for files from plugins?

File: app/models/smart_proxy.rb (hunk 11 above)
Comment (shlomizadok) [inline app/models/smart_proxy.rb]: Yeah, as it will minimum have the Version status

File: app/services/proxy_status/base.rb [hunk 13]
```diff
@@ -0,0 +1,39 @@
+module ProxyStatus
+  class Base
+    def initialize(proxy, opts = {}, api_class_name = 'Version')
+      @proxy = proxy
+      @cache_duration = opts[:cache_duration] || 3.minutes
+      @api_class_name = api_class_name
+    end
+
+    def api
```
Comment (ohadlevy) [inline app/services/proxy_status/base.rb]: imho should be private.

File: app/services/proxy_status/base.rb [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+      @api = "ProxyAPI::#{api_class_name}".classify.constantize.new(:url => proxy.url)
+    end
+    def revoke_cache!
```
Comment (ohadlevy) [inline app/services/proxy_status/base.rb:9]: nitpick: missing space

File: app/models/smart_proxy.rb [hunk 15]
```diff
@@ -88,12 +86,19 @@ def has_feature?(feature)
     self.features.any? { |proxy_feature| proxy_feature.name == feature }
   end
 
-  private
-
   def proxy_status
-    @proxy_status ||= ProxyStatus.new(self)
+    @proxy_status ||= { :version => ProxyStatus::Version.new(self) }
+    features.map(&:name).each do |feature|
+      normalized_feature_name = feature.delete(' ')
```
Comment (ohadlevy) [inline app/models/smart_proxy.rb]: do you want or not to chace the proxy status? if you do, then this is not cache, but rather recalculated every time.

Comment (ohadlevy) [inline app/models/smart_proxy.rb]: not sure what happened to my original comment

File: app/models/smart_proxy.rb [hunk 16]
```diff
@@ -88,12 +86,19 @@ def has_feature?(feature)
...
+      status = ProxyStatus.find_status_by_humanized_name(normalized_feature_name)
```
Comment (ohadlevy) [inline app/models/smart_proxy.rb]: ditto about downcase, normalized name should already return down case imho

File: app/services/proxy_status.rb (hunk 12 above)
Comment (ares) [inline app/services/proxy_status.rb:14]: yes we do, plugins can add require_dependency in `config.to_prepare` block (as they normally do with any other STI)

OTOH if you add some API for registering custom status similar to `register_custom_status(RandomStatus)` that would be called in `config.to_prepare`, it shouldn't be needed since calling RandomStatus should autoload the class

File: app/models/smart_proxy.rb (hunk 16 above)
Comment (shlomizadok) [inline app/models/smart_proxy.rb]: Normalized name can not be downcased, as it is passed later as a class name. I can downcase it here and the upcase it on `ProxyStatus::Base` but it seems un-necessary

File: app/models/smart_proxy.rb (hunk 15 above)
Comment (shlomizadok) [inline app/models/smart_proxy.rb]: Why isn't it cached?

File: app/models/smart_proxy.rb [hunk 17]
```diff
@@ -88,12 +86,19 @@ def has_feature?(feature)
     self.features.any? { |proxy_feature| proxy_feature.name == feature }
   end
 
-  private
-
   def proxy_status
-    @proxy_status ||= ProxyStatus.new(self)
+    @proxy_status ||= { :version => ProxyStatus::Version.new(self) }
+    features.map(&:name).each do |feature|
```
Comment (ohadlevy) [inline app/models/smart_proxy.rb]: you rely on name uniqueness, please ensure we enforce it on the feature model.

File: app/models/smart_proxy.rb (hunk 15 above)
Comment (ohadlevy) [inline app/models/smart_proxy.rb]: because you map over the features regardless of the @proxy_status object?

File: app/models/smart_proxy.rb (hunk 16 above)
Comment (ohadlevy) [inline app/models/smart_proxy.rb]: ok.

File: app/services/proxy_status/base.rb [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+    attr_reader :proxy, :cache_duration, :api_class_name
+
+    def api
+      @api = "ProxyAPI::#{api_class_name}".classify.constantize.new(:url => proxy.url)
+    end
+
+    def fetch_proxy_data
+      begin
+        yield
+      rescue *ProxyStatus::CONNECTION_ERRORS => exception
```
Comment (ohadlevy) [inline app/services/proxy_status/base.rb]: nitpick: you can drop the begin / end and simple rescue the entire method.

File: app/services/proxy_status/tftp.rb [hunk 19]
```diff
@@ -0,0 +1,24 @@
+module ProxyStatus
+  class TFTP < Base
+    def initialize(proxy, opts = {})
+      super(proxy, opts, 'TFTP')
+    end
+
+    def tftp_server
+      Rails.cache.fetch("proxy_#{proxy.id}/tftp_server", :expires_in => cache_duration) do
```
Comment (ohadlevy) [inline app/services/proxy_status/tftp.rb]: can we extend the fetch_proxy_data to accept opts and then you can define if you want to cache or not? this pattern is going to not DRY alot

File: app/controllers/smart_proxies_controller.rb (hunk 8 above)
Comment (dLobatog) [inline app/controllers/smart_proxies_controller.rb]: I was going to suggest this exact same thing on the original PR but didn't want to delay it any more. If you could do it it'd be great. Basically it'd be similar to the way `fetch_proxy_data` works. Your external calls would look like

```
requested_data do 
  @proxy_status[:tftp].server
end
```

I'd also suggest changing the name `requested_data` (generic and meaningless) to something more specific, even `fetch_proxy_data` again if you want?

File: app/models/smart_proxy.rb (hunk 17 above)
Comment (dLobatog) [inline app/models/smart_proxy.rb]: Also if you can extract the features initialization to a method that'd be great.

File: app/models/smart_proxy.rb [hunk 20]
```diff
@@ -88,12 +86,19 @@ def has_feature?(feature)
     self.features.any? { |proxy_feature| proxy_feature.name == feature }
   end
 
-  private
-
   def proxy_status
-    @proxy_status ||= ProxyStatus.new(self)
+    @proxy_status ||= { :version => ProxyStatus::Version.new(self) }
```
Comment (dLobatog) [inline app/models/smart_proxy.rb]: Nitpick, we're already in SmartProxy, so `@status` sounds a bit less redundant?

File: app/services/proxy_status/base.rb [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+      @api_class_name = api_class_name
+    end
+
+    def revoke_cache!
+      # As memcached does not support delete_matched, we need to delete each
+      Rails.cache.delete(cache_key)
+    end
+
+    def cache_key
+      raise NotImplementedError _('Implement in subclasses')
```
Comment (dLobatog) [inline app/services/proxy_status/base.rb]: This is missing a comma between Error and the string, Also the string should say something like `cache_key not implemented for ProxyStatus#{self.class}` or something similar, so that it stands out on logs

File: app/services/proxy_status/tftp.rb [hunk 22]
```diff
@@ -0,0 +1,24 @@
+module ProxyStatus
+  class TFTP < Base
+    def initialize(proxy, opts = {})
+      super(proxy, opts, 'TFTP')
+    end
+
+    def tftp_server
```
Comment (dLobatog) [inline app/services/proxy_status/tftp.rb]: Nitpick, how about just `server`? Think about how it is called, `proxy_status[:tftp].tftp_server` vs `proxy_status[:tftp].server`

File: test/unit/proxy_status_test.rb [hunk 23]
```diff
@@ -7,27 +7,34 @@ class ProxyStatusTest < ActiveSupport::TestCase
   end
 
   test '#versions_cache_key' do
-    proxy_status = ProxyStatus.new(@proxy)
-    assert_equal("proxy_#{@proxy.id}/versions", proxy_status.versions_cache_key)
+    proxy_status = ProxyStatus::Version.new(@proxy)
```
Comment (dLobatog) [inline test/unit/proxy_status_test.rb]: `mkdir test/unit/proxy_statuses` and test each one independently?

**Comment by dLobatog** [comment]
@shlomizadok Almost ready in my opinion, great work! 
The things that are missing look easy to fix. Let's summarize:
- [x] [Rewrite `requested_data` to perform like `fetch_proxy_data`](https://github.com/theforeman/foreman/pull/3069#discussion_r49733746)
- [x] [Validate feature name uniqueness](https://github.com/theforeman/foreman/pull/3069/files#r49969534)
- [ ] Extract function: [initialize_features_status](https://github.com/theforeman/foreman/pull/3069/files#r50024355)
- [x] [Rename proxy_status to just status under smart_proxy](https://github.com/theforeman/foreman/pull/3069#discussion_r50024411)
- [x] [Rename tftp_server to just server under tftp status](https://github.com/theforeman/foreman/pull/3069#discussion_r50025154)
- [x] Fix [typo on cache_key](https://github.com/theforeman/foreman/pull/3069/files#r50024959)
- [x] [rescue entire `fetch_proxy_data` method](https://github.com/theforeman/foreman/pull/3069/files#r49969619)
- [x] Make `fetch_proxy_data` [accept option to cache or not](https://github.com/theforeman/foreman/pull/3069/files#r49969675)
- [x] Test each ProxyStatus base, tftp and version in their own files, under test/unit/proxy_statuses.

I'd call this ready to merge after that list is merged, am I overlooking anything you saw before, @ohadlevy ?

**Comment by shlomizadok** [comment]
@dLobatog - Thanks. Each item should be checked

File: app/models/feature.rb [hunk 24]
```diff
@@ -3,7 +3,7 @@ class Feature < ActiveRecord::Base
   friendly_id :name
   has_and_belongs_to_many :smart_proxies
   validates_lengths_from_database
-  validates :name, :presence => true
+  validates :name, :presence => true, :uniqueness => true
```
Comment (ohadlevy) [inline app/models/feature.rb]: consider adding { case_sensitive: false }.

File: app/controllers/smart_proxies_controller.rb [hunk 25]
```diff
@@ -63,9 +72,12 @@ def destroy
 
   private
 
-  def requested_data(data_name)
-    data = @smart_proxy.public_send(data_name)
-    render :json => {:success => true, :message => data}
+  def find_status
+    @proxy_status = @smart_proxy.proxy_status
```
Comment (ohadlevy) [inline app/controllers/smart_proxies_controller.rb]: isnt that now .status?

File: app/models/smart_proxy.rb [hunk 26]
```diff
@@ -88,12 +86,19 @@ def has_feature?(feature)
     self.features.any? { |proxy_feature| proxy_feature.name == feature }
   end
 
-  private
-
-  def proxy_status
-    @proxy_status ||= ProxyStatus.new(self)
+  def status
+    @status ||= { :version => ProxyStatus::Version.new(self) }
+    features.map(&:name).each do |feature|
```
Comment (ohadlevy) [inline app/models/smart_proxy.rb]: I'm not following why you do that (or why github keeps removing my comment about this).

File: app/services/proxy_status.rb (hunk 12 above)
Comment (ohadlevy) [inline app/services/proxy_status.rb:14]: /cc @ShimShtein

File: app/services/proxy_status/base.rb [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+  class Base
+    def initialize(proxy, opts = {}, api_class_name = 'Version')
+      @proxy = proxy
+      @cache_duration = opts[:cache_duration] || 3.minutes
+      @cache_status = opts[:cache_status].nil? ? true : opts[:cache_status]
+      @api_class_name = api_class_name
+    end
+
+    def revoke_cache!
+      # As memcached does not support delete_matched, we need to delete each
```
Comment (ohadlevy) [inline app/services/proxy_status/base.rb]: remove comment?

File: app/services/proxy_status/base.rb [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+      Rails.cache.delete(cache_key)
+    end
+
+    def cache_key
+      raise NotImplementedError, _("cache_key not implemented for ProxyStatus::#{self.class}")
+    end
+
+    def self.humanized_name
+      # Should be named as the proxy feature name!
```
Comment (ohadlevy) [inline app/services/proxy_status/base.rb]: does it make sense to test if the feature exists ?

Comment (ohadlevy) [inline app/services/proxy_status/base.rb]: either here or at the status registry time?

File: app/services/proxy_status/tftp.rb [hunk 29]
```diff
@@ -0,0 +1,22 @@
+module ProxyStatus
+  class TFTP < Base
+    def initialize(proxy, opts = {})
+      super(proxy, opts, 'TFTP')
```
Comment (ohadlevy) [inline app/services/proxy_status/tftp.rb]: nitpick, can probably use self.humanized_name by default instead of override the initalizer.

File: test/unit/proxy_statuses/proxy_status_tftp.rb [hunk 30]
```diff
@@ -0,0 +1,42 @@
+require 'test_helper'
+
+class ProxyStatusTftp < ActiveSupport::TestCase
```
Comment (dLobatog) [inline test/unit/proxy_statuses/proxy_status_tftp.rb]: This test is not being run, the file needs to end in `_test`, and the class has to be named `ProxyStatusTftpTest`

File: test/unit/proxy_statuses/proxy_status_versions.rb [hunk 31]
```diff
@@ -0,0 +1,34 @@
+require 'test_helper'
+
+class ProxyStatusVersions < ActiveSupport::TestCase
```
Comment (dLobatog) [inline test/unit/proxy_statuses/proxy_status_versions.rb]: Same as Tftp, this test doesn't run because of the file & name class lack _test, *Test

File: app/services/proxy_status/base.rb [hunk 32]
```diff
@@ -0,0 +1,45 @@
...
+      @api_class_name = api_class_name
+    end
+
+    def revoke_cache!
+      # As memcached does not support delete_matched, we need to delete each
+      Rails.cache.delete(cache_key)
+    end
+
+    def cache_key
+      raise NotImplementedError, _("cache_key not implemented for ProxyStatus::#{self.class}")
```
Comment (dLobatog) [inline app/services/proxy_status/base.rb]: This is not translator friendly, but I guess it's fine since only developers should ever see this exception?

File: app/services/proxy_status/base.rb (hunk 28 above)
Comment (dLobatog) [inline app/services/proxy_status/base.rb]: I'd be fine with this coming in a separate PR, but yeah, that'd be nice.

**Comment by dLobatog** [comment]
@shlomizadok Almost there, there were a few things missing:
- [x] Validate feature name uniqueness (case insensitive)
- [ ] Extract function: [initialize_features_status](https://github.com/theforeman/foreman/pull/3069/files#r50024355)
- [x] Make sure all test files finish on `_test` and the class names on 'Test' 
- [x] Controller should call proxy.status not .proxy_status - https://github.com/theforeman/foreman/pull/3069#discussion_r50082648

**Comment by shlomizadok** [comment]
fixed

**Comment by dLobatog** [comment]
@shlomizadok that was quick :smile: hopefully it's 100% done now!

**Comment by shlomizadok** [comment]
Probably not :/

File: app/models/smart_proxy.rb [hunk 33]
```diff
@@ -88,12 +86,20 @@ def has_feature?(feature)
...
   end
 
-  private
-
-  def proxy_status
-    @proxy_status ||= ProxyStatus.new(self)
+  def status
+    return @status if @status
+    @status = { :version => ProxyStatus::Version.new(self) }
+    features.map(&:name).each do |feature|
```
Comment (dLobatog) [inline app/models/smart_proxy.rb]: @ohadlevy I know it's infuriating, but your comment got removed again?

File: test/unit/feature_test.rb [hunk 34]
```diff
@@ -3,4 +3,5 @@
 class FeatureTest < ActiveSupport::TestCase
   should have_and_belong_to_many(:smart_proxies)
   should validate_presence_of(:name)
+  should validate_uniqueness_of(:name)
```
Comment (dLobatog) [inline test/unit/feature_test.rb]: `validate_uniqueness_of(:name).case_insensitive`

File: app/models/smart_proxy.rb [hunk 35]
```diff
@@ -88,12 +86,20 @@ def has_feature?(feature)
...
+      normalized_feature_name = feature.delete(' ')
+      status = ProxyStatus.find_status_by_humanized_name(normalized_feature_name)
```
Comment (dLobatog) [inline app/models/smart_proxy.rb]: nitpick, 'status' here as a local variable & as a instance variable `@status` can be confusing to newcomers, perhaps `feature_status` is a little less confusing. `status_symbol` would then become `feature_status_symbol`

**Comment by dLobatog** [comment]
@shlomizadok Actually yeah :smile: My only last comment is that nitpick about `status` and `@status`, and to fix the feature validation test to ensure it's case insensitive. RTM after that & answering to @ohadlevy 's comment https://github.com/theforeman/foreman/pull/3069#discussion_r49968973

File: app/models/smart_proxy.rb (hunk 33 above)
Comment (dLobatog) [inline app/models/smart_proxy.rb]: https://github.com/theforeman/foreman/pull/3069#discussion_r49968973

File: app/models/smart_proxy.rb (hunk 15 above)
Comment (shlomizadok) [inline app/models/smart_proxy.rb]: Changed

**Comment by shlomizadok** [comment]
Changed variable names. Changed feature_test, Changed status caching on https://github.com/theforeman/foreman/pull/3069#discussion_r49968973

File: app/models/smart_proxy.rb (hunk 33 above)
Comment (shlomizadok) [inline app/models/smart_proxy.rb]: Is this code properly sets the caching?

``` ruby
def status
  return @status if @status
  @status = { :version => ProxyStatus::Version.new(self) }
  features.map(&:name).each do |feature|
    normalized_feature_name = feature.delete(' ')
    feature_status = ProxyStatus.find_status_by_humanized_name(normalized_feature_name)
    feature_status_symbol = normalized_feature_name.downcase.to_sym
    @status[feature_status_symbol] = feature_status.new(self) if feature_status
  end
  @status
end
```

@ohadlevy @dLobatog

**Comment by dLobatog** [comment]
@shlomizadok Outstanding! :smile: I'll wait for Jenkins to finish, & if possible for @ohadlevy to ack that https://github.com/theforeman/foreman/pull/3069/files#diff-a33b3d60342842d2113ddbb8eb1281a5R91 (cache is only revoked when the users clicks on refresh? I'm fine with that, just checking), then :bomb: we can merge it. I'll test it once again if @ohadlevy acks that change but I didn't find anything broken last time.

File: app/models/smart_proxy.rb (hunk 33 above)
Comment (ohadlevy) [inline app/models/smart_proxy.rb]: @shlomizadok I would write it like this:

``` ruby
def statuses
  return @statuses if @statuses

  @status = {}
  features.each do |feature|
    name = feature.name.delete(' ')
    if (status = ProxyStatus.find_status_by_humanized_name(name))
      @statuses[name.to_sym] = status.new(self)
    end
  end
  @statuses[:version] = ProxyStatus::Version.new(self) }

  @statuses
end


```

Comment (ohadlevy) [inline app/models/smart_proxy.rb]: note pluralized status

File: app/services/proxy_status/base.rb [hunk 36]
```diff
@@ -0,0 +1,43 @@
+module ProxyStatus
+  class Base
+    def initialize(proxy, opts = {})
+      @proxy = proxy
+      @cache_duration = opts[:cache_duration] || 3.minutes
+      @cache_status = opts[:cache_status].nil? ? true : opts[:cache_status]
```
Comment (ohadlevy) [inline app/services/proxy_status/base.rb]: nitpick: why call it cache_status? isnt it cache true / false ?

File: app/services/proxy_status/base.rb [hunk 37]
```diff
@@ -0,0 +1,43 @@
...
+      # Should be named as the proxy feature name!
+      'Base'
+    end
+
+    private
+
+    attr_reader :proxy, :cache_duration, :api_class_name, :cache_status
+
+    def api
+      @api = "ProxyAPI::#{self.class.humanized_name}".classify.constantize.new(:url => proxy.url)
```
Comment (ohadlevy) [inline app/services/proxy_status/base.rb:26]: what happens if no such class exists? afaiu this will be reported in the UI as unable to initialize class name or something like that.
I wonder if we should have a nicer error or its fine as its developer oriented?

**Comment by shlomizadok** [comment]
@dLobatog @ohadlevy - removed uniqueness from feature, as it breaks all test (to be precise it broke 395 / 1600 [which seems reasonable, actually...])
Anyhow, the problem is that in the tests we always do `FactoryGirl.create(:feature, :feature_name)` and since FactoryGirl does not support `first_or_create` - refactoring this needs to be on a separate PR IMO.

File: app/models/smart_proxy.rb [hunk 38]
```diff
@@ -88,12 +86,20 @@ def has_feature?(feature)
     self.features.any? { |proxy_feature| proxy_feature.name == feature }
   end
 
-  private
+  def status
```
Comment (ohadlevy) [inline app/models/smart_proxy.rb]: since github ate my comment again, I'll paste it here again:

``` ruby
def statuses
  return @statuses if @statuses

  @status = {}
  features.each do |feature|
    name = feature.name.delete(' ')
    if (status = ProxyStatus.find_status_by_humanized_name(name))
      @statuses[name.to_sym] = status.new(self)
    end
  end
  @statuses[:version] = ProxyStatus::Version.new(self) }

  @statuses
end
```

not the plural statuses

File: app/services/proxy_status/base.rb (hunk 36 above)
Comment (ohadlevy) [inline app/services/proxy_status/base.rb]: unresolved.

File: app/services/proxy_status/tftp.rb [hunk 39]
```diff
@@ -0,0 +1,18 @@
+module ProxyStatus
+  class TFTP < Base
+    def server
+      fetch_proxy_data do
+        api.bootServer
+      end
+    end
+
+    def cache_key
+      "proxy_#{proxy.id}/tftp_server"
```
Comment (ohadlevy) [inline app/services/proxy_status/tftp.rb]: it sounds like the cache key could default to:

``` ruby
def cache_key
  "proxy"_#{proxy.id}/#{self.humanized_name}"
end
```

**Comment by ohadlevy** [comment]
@dLobatog i'm fine with the cache behaviour, after all it only caches for 3 minutes anyway. refresh make sure the cache is invalidated.

**Comment by ohadlevy** [comment]
@shlomizadok ui needs to change from refresh features to just refresh / reload imho.

File: app/models/smart_proxy.rb (hunk 38 above)
Comment (shlomizadok) [inline app/models/smart_proxy.rb]: https://github.com/theforeman/foreman/pull/3069/files#diff-a33b3d60342842d2113ddbb8eb1281a5R89

**Comment by shlomizadok** [comment]
Changed.
Can this be reviewed once again, please? Thanks

**Comment by dLobatog** [comment]
:+1:  if tests pass

**Comment by dLobatog** [comment]
[test]

File: test/unit/proxy_statuses/proxy_status_test.rb [hunk 40]
```diff
@@ -0,0 +1,13 @@
...
+
+class ProxyStatusTest < ActiveSupport::TestCase
+  setup do
+    @proxy = FactoryGirl.
+      build_stubbed(:template_smart_proxy, :url => 'https://secure.proxy:4568')
+  end
+
+  test '#versions_cache_key' do
+    proxy_status = ProxyStatus::Version.new(@proxy)
+    assert_equal("proxy_#{@proxy.id}/versions", proxy_status.cache_key)
```
Comment (dLobatog) [inline test/unit/proxy_statuses/proxy_status_test.rb]: The cache key should be 'proxy_id/Version' - http://ci.theforeman.org/job/test_develop_pr_core/7801/database=mysql,label=fast,ruby=2.2/testReport/junit/%28root%29/ProxyStatusTest/test_0001__versions_cache_key/

**Comment by dLobatog** [comment]
Merged as 37edf55caf856da2c53628f65dcfb40c2f3ac18b, thanks @shlomizadok!
